File: Algorithms/1200. Minimum Absolute Difference/php/solution.counting-sort.php
class Solution {
    /**
     * @param Integer[] $arr
     * @return Integer[][]
     */
    function minimumAbsDifference($arr) {
        $minElement = min($arr);
        $maxElement = max($arr);
        $shift = -$minElement;
        $line = array_fill(0, $maxElement - $minElement + 1, 0);
        $answer = array();
        
        // 標記數字出現
        foreach ($arr as $num) {
            $line[$num + $shift] = 1;
        }
        
        $minPairDiff = $maxElement - $minElement;
        $prev = 0;
        
        // 遍歷找出最小差值對
        for ($curr = 1; $curr <= $maxElement + $shift; $curr++) {
            if ($line[$curr] === 0) continue;
            
            if ($curr - $prev === $minPairDiff) {
                array_push($answer, [$prev - $shift, $curr - $shift]);
            } elseif ($curr - $prev < $minPairDiff) {
                $answer = array();
                $minPairDiff = $curr - $prev;
                array_push($answer, [$prev - $shift, $curr - $shift]);
            }
            $prev = $curr;
        }
        
        return $answer;
    }
}

File: Algorithms/1200. Minimum Absolute Difference/swift/solution.counting-sort.swift
class Solution {
    func minimumAbsDifference(_ arr: [Int]) -> [[Int]] {
        let minElement = arr.min()!
        let maxElement = arr.max()!
        let shift = -minElement
        var line = Array(repeating: false, count: maxElement - minElement + 1)
        var answer: [[Int]] = []
        
        // 標記數字出現
        for num in arr {
            line[num + shift] = true
        }
        
        var minPairDiff = maxElement - minElement
        var prev = 0
        
        // 遍歷找出最小差值對
        for curr in 1...maxElement + shift {
            guard line[curr] else { continue }
            
            if curr - prev == minPairDiff {
                answer.append([prev - shift, curr - shift])
            } else if curr - prev < minPairDiff {
                answer = [[prev - shift, curr - shift]]
                minPairDiff = curr - prev
            }
            prev = curr
        }
        
        return answer
    }
}
